### PR TITLE
use timeouts when sending messages for stream open, close, and reset.

### LINKED ISSUE #51: stalled writer goroutines
Grepping through goroutine traces of our relay infrastructure, I found stalled goroutines:

```
goroutine 2826624695 [select, 3 minutes]:
github.com/libp2p/go-mplex.(*Multiplex).sendMsg(0xc1f430fc00, 0xf77f00, 0xc0000cc010, 0x18350, 0xc3aefa9990, 0x5, 0x20, 0x0, 0x0)
#011/home/ubuntu/go/pkg/mod/github.com/vyzo/go-mplex@v0.0.0-20190506084127-53d5bb35e24a/multiplex.go:143 +0xf5
github.com/libp2p/go-mplex.(*Multiplex).NewNamedStream(0xc1f430fc00, 0xc044180008, 0x5, 0xc3aefa9a20, 0xe8ec18, 0xc1720a67a0)
#011/home/ubuntu/go/pkg/mod/github.com/vyzo/go-mplex@v0.0.0-20190506084127-53d5bb35e24a/multiplex.go:217 +0x1a0
github.com/libp2p/go-mplex.(*Multiplex).NewStream(...)
[...]
```

When we write control messages, we transfer the deadline from the context to the connection, but the `sendMsg` method is always called with the background context, which never carries a deadline. So if one write stalls, it'll stall forever and it'll also block any subsequent sending of control messages on that multiplexed connection.

See: https://github.com/libp2p/go-mplex/blob/master/multiplex.go#L152

### PR DESCRIPTION
**Pull request by vyzo**
Closes #51 #48

### CODE REVIEW

**Comment by vyzo** [comment]
test failures are #47

File: multiplex.go [hunk 1]
```diff
@@ -214,7 +223,10 @@ func (mp *Multiplex) NewNamedStream(name string) (*Stream, error) {
 	mp.channels[s.id] = s
 	mp.chLock.Unlock()
 
-	err := mp.sendMsg(context.Background(), header, []byte(name))
+	ctx, cancel := context.WithTimeout(bgCtx, time.Minute)
```
Comment (raulk) [inline multiplex.go]: Can we move these timeouts to a `SocketGracePeriod` var, so they can be adjusted?

File: multiplex.go [hunk 2]
```diff
@@ -62,6 +64,13 @@ type Multiplex struct {
 	chLock   sync.Mutex
 }
 
+func init() {
+	// this context is here to allow us to attach deadlines to messages without
+	// spawning a new goroutine for propagateCancel each time.
+	// we ignore the cancel function because it will never get canceled.
+	bgCtx, _ = context.WithCancel(context.Background())
```
Comment (raulk) [inline multiplex.go]: Do we actually need to create a new context here? Why aren't we able to use `context.Background()` directly?

**Comment by vyzo** [comment]
We might also consider _not_ sending a reset message at all, just swallow the remote error.

File: multiplex.go (hunk 2 above)
Comment (vyzo) [reply multiplex.go]: I think it will have `propagateCancel` goroutines because it's not hooked to a cancellation.

File: multiplex.go (hunk 1 above)
Comment (vyzo) [reply multiplex.go]: Sure will do. Do we want separate timeouts for reset and stream open or just use a single one?

Comment (vyzo) [reply multiplex.go]: Done.

File: multiplex.go (hunk 2 above)
Comment (raulk) [reply multiplex.go]: It shouldn't, as this guard would short circuit the propagation: https://github.com/golang/go/blob/master/src/context/context.go#L245

Comment (vyzo) [reply multiplex.go]: ok, will get rid of it then.

Comment (vyzo) [reply multiplex.go]: done

**Comment by vyzo** [comment]
summoning @Stebalien @raulk -- this has been thoroughly tested, let's review and get it merged!

**Pull request merged**